### PR TITLE
Feature/sas

### LINKED ISSUE #3047: E2E: SAS Verification (app)
@lampholder commented on [Fri Jan 04 2019](https://github.com/vector-im/riot-web/issues/8021)

Current web flow:
[Verif_emoji.pdf](https://github.com/vector-im/riot-android/files/2961905/Verif_emoji.pdf)

This comprises the SAS components of @nadonomy's Managing Trust wireframes (inbound and outbound).

So that's:
- outbound:
    - initiate a request to verify a user via UX in the member info panel
    - poke the js sdk to broadcast that request to all of that user's devices
    - display a waiting screen whilst we wait for the user to accept the request
    - show the verification hex string and a 'confirm' button when the request has been accepted
    - complete the verification when the confirm button is clicked
- inbound:
    - display an inbound request to verify on all of your devices (**if the request is not too old**)
    - expire all the other devices' request to verify when you accept on one device
    - show the verification hex string on your device
    - show that verification is complete once the verifier has clicked confirm (is this right?)

---

@jryans commented on [Wed Jan 30 2019](https://github.com/vector-im/riot-web/issues/8021#issuecomment-458984196)

Fixed by https://github.com/matrix-org/matrix-react-sdk/pull/2461.

### PR DESCRIPTION
**Pull request by BillCarsonFr**
* [x] Start verification request emoji and/or decimal
* [x] Handle incoming verification request emoji and/or decimal
* [x] Fine tuning of cancelled request (interrupted/cancelled by me / by other)
* [x] Handle activity return code to refresh calling screens after verification
* [x] Stack incoming request and process one by one
* [x] Code Cleaning
* [x] Changes has been tested on an Android device or Android emulator with API 16
* [x] UI change has been tested on both light and dark themes
* [x] Pull request is based on the develop branch
* [x] Pull request updates [CHANGES.rst](https://github.com/vector-im/riot-android/blob/develop/CHANGES.rst)
* [x] Pull request includes screenshots or videos if containing UI changes
* [x] Perform complete integration tests with riot-web (app & develop)
* [x] Perform complete integration tests with other riot-android (Jelly bean- decimal- to Pie)
* [x] Perform complete integration tests with riot-ios

fixes #3047

[Screenshots](https://github.com/vector-im/riot-android/files/3047360/sas_droid.pdf)
sas_emoji
sas_code

### CODE REVIEW

File: vector/src/main/res/layout/fragment_sas_verification_incoming_request.xml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        android:layout_width="40dp"
+        android:layout_height="40dp"
+        android:layout_marginTop="@dimen/layout_vertical_margin"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/sas_incoming_request_title"
+        tools:src="@tools:sample/avatars" />
+
+    <TextView
+        android:id="@+id/sas_incoming_request_user_id"
```
Comment (manuroe) [inline vector/src/main/res/layout/fragment_sas_verification_incoming_request.xml]: We should display user display name

Comment (bmarty) [reply vector/src/main/res/layout/fragment_sas_verification_incoming_request.xml]: done

File: vector/src/main/res/layout/fragment_sas_verification_verified.xml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        <TextView
+            android:id="@+id/sas_incoming_request_title"
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="20dp"
+            android:layout_marginTop="30dp"
+            android:layout_marginEnd="20dp"
+            android:layout_marginBottom="16dp"
+            android:gravity="center"
+            android:text="@string/sas_incoming_request_title"
```
Comment (manuroe) [inline vector/src/main/res/layout/fragment_sas_verification_verified.xml]: This `sas_incoming_request_title` is also displayed for outgoing device verification request.

Comment (bmarty) [reply vector/src/main/res/layout/fragment_sas_verification_verified.xml]: Good point, done

**Comment by bmarty** [comment]
I've finished the review @BillCarsonFr

**Pull request merged**